### DISCUSSION TITLE
Replace image blob

### DISCUSSION THREAD

**Discussion by micha149** [newren/git-filter-repo#321] [Q&A]
Hi.

Some time ago one committed a quite large image to our git repository. Now we are splitting this repository into multiple smaller repos and I want to take the opportunity to replace this big images blob with a smaller one.

Is there a simple way to do this? The only way I can image from the docs is to write a python callback.

**Comment by newren** [comment]
You can do it with a python callback, but I suspect the easiest way would be to get git's hashes for both the old and new blob and then run:

```
git replace -f $old_blob_id $new_blob_id
```

followed by running filter-repo.  filter-repo pays attention to replace refs and makes them permanent automatically, no special flags required.  (Although, if you were only replacing the image and not rewriting anything else or splitting the repository, you'd have to pass --force to filter-repo, because filter-repo has a safety check to error out when not given any filtering options.)

**Comment by mrpeverill** [comment]
I was trying to do something very similar (in my case, substituting a one pdf with another that had had a minor change made), but I was not able to get the above git replace command to work. My steps were to:

1. Make a fresh mirrored clone
2. use 'git hash-object -w --path=$directory_of_file_to_be_replaced $Location_of_modified_file #(get the new hash)
3. git replace -f $old_hash $new_hash
4. git replace --format=long #to confirm
5. git-filter-repo --force

This would give the message:
`warning: refs/replace/$old_hash: Unexpected object of type blob, skipping.
`

Instead, I followed steps 1 and 2 above, then ran the following:
`git-filter-repo --force --commit-callback '
  for change in commit.file_changes:
    if change.blob_id == b"$old_hash":
      change.blob_id = b"$new_hash"
  '`

I don't really know what I'm doing and probably there is something basic I was missing with the 'git replace' option, but this worked for me.
